BUG: set_index() and reset_index() not preserving object dtypes

trevorbye opened the pull request:
* changes to `set_index()` and `ensure_index_from_sequences()` to pass current dtype of key through call stack onto Index constructor, rather than letting Index constructor infer the dtype. For now I just aimed to patch a single key param from `set_index()`, as that is the lowest hanging fruit for this issue I think.

* removed `maybe_convert_objects()` call from `_maybe_casted_values()`. This will preserve an `object` dtype on a `set_index().reset_index()` round-trip. I came to the exact same conclusion as the user in [this PR comment](https://github.com/pandas-dev/pandas/pull/27370#issuecomment-511051996), for context. This makes sense conceptually, if your index was an object dtype, it should be an object dtype as a column. Of course, this is somewhat subjective so let me know your opinions, but this is what I interpreted from the discussion on the git issue.

* both of these changes break quite a few existing tests, most of which are testing specifically for the exact opposite conditions that this PR is intended to patch **e.g.** tests that look for object dtypes being coerced to more-specific types. What's the process for handling failed tests for a change like this? Would you attempt to refactor all, or delete the ones where the logic is no longer something to test for? I also ran `pytest pandas` on a fresh pull of master and had 13 fails, with 108 failures on my branch specifically, so I am also curious what to do about those previously-failing tests, or what is acceptable for a merge.

- [x] closes #30517 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
initial test case

Commit:
test change

Commit:
changes to set_index to preserve dtype from field

Commit:
fixing test error

Commit:
removed object type inf from _maybe_casted_values

Commit:
missing char

pep8speaks commented:
Hello @trevorbye! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/76c0f8a864924500046ec9a3cc3ff5267818480b/pandas/core/frame.py):

> [Line 4499:1](https://github.com/pandas-dev/pandas/blob/76c0f8a864924500046ec9a3cc3ff5267818480b/pandas/core/frame.py#L4499): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace



* In the file [`pandas/tests/frame/test_set_index_preserve_dtype.py`](https://github.com/pandas-dev/pandas/blob/76c0f8a864924500046ec9a3cc3ff5267818480b/pandas/tests/frame/test_set_index_preserve_dtype.py):

> [Line 32:89](https://github.com/pandas-dev/pandas/blob/76c0f8a864924500046ec9a3cc3ff5267818480b/pandas/tests/frame/test_set_index_preserve_dtype.py#L32): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 88 characters)
> [Line 33:89](https://github.com/pandas-dev/pandas/blob/76c0f8a864924500046ec9a3cc3ff5267818480b/pandas/tests/frame/test_set_index_preserve_dtype.py#L33): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 88 characters)
> [Line 33:97](https://github.com/pandas-dev/pandas/blob/76c0f8a864924500046ec9a3cc3ff5267818480b/pandas/tests/frame/test_set_index_preserve_dtype.py#L33): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

trevorbye commented:
@jreback for visibility

jreback commented on pandas/tests/frame/test_set_index_preserve_dtype.py:
```diff
@@ -0,0 +1,40 @@
+
+
+@pytest.mark.parametrize('test_dtype', [object, 'int64'])
+def test_dtypes(test_dtype):
```
don’t make a new file we have loads of tests in alter_axes u can just add it therr

trevorbye replied:
Ok will do

jreback reviewed (changes requested):
try to make an even smaller change this likely will break lots of things

trevorbye commented:
> try to make an even smaller change this likely will break lots of things

Any suggestions for making these changes smaller? I tried to touch as little as possible, but not sure how I could slim it down further.

Or another question would be, what do you see breaking a lot? 

Thanks

Commit:
moved tests

trevorbye commented:
@jreback @TomAugspurger any suggestions how to make this change smaller per your comment?

What are your thoughts on implementing this as non-default behavior? For example, adding a new optional param to both `set_index()` and `reset_index()` something like a boolean for `preserve_dtype`, that would enable the changes I've made here rather than make them the default. This would stop all these current tests from failing, and not break user's code. The tradeoff is the user would have to specify this param to get the new behavior, so I suppose that is a design question.

From the conversation in the [issue](https://github.com/pandas-dev/pandas/issues/30517) I had the impression it should be the new default behavior, but I could have misinterpreted. Either way let me know what you think the path forward is and I will work on it. Thanks

TomAugspurger commented:
FYI @trevorbye you've commited some pickle files that are being left around by a some test (we haven't discovered which one yet).

I not sure what Jeff had in mind for making this smaller.

I think round-tripping the dtypes should be the default behavior. It not doing so today is a bug.

In theory, we could handle this in the Index constructor and just not have it infer the dtype when given a typed object. But that would probably be an API breaking change.

trevorbye commented:
@TomAugspurger thanks, I'll get rid of those files. 

What is your best practice for refactoring/removing tests for a breaking change like this? There are quite a few failing, would you change the logic to reflect the new type inference (or lack thereof, in this case)? 

For Index constructor, if you call it yourself you can pass the dtype param which will disable the type coercion. But, when the constructor is called from **inside** `set_index()`, it doesn't pass dtype param. The change I made there simply grabs the type of the col you're trying to create an index from, and passes that on to the constructor and uses the logic already in place. 

Still learning the code base, but I'm wondering if there's a way to do this with less overall impact than the way I did it.

TomAugspurger commented:
I'm not sure. But if we're calling the dtype not being preserved a bug,
then we'll just need to update the tests that rely on the constructors
re-inferring dtypes.

On Mon, Jan 13, 2020 at 10:19 PM trevorbye <notifications@github.com> wrote:

> @TomAugspurger <https://github.com/TomAugspurger> thanks, I'll get rid of
> those files.
>
> What is your best practice for refactoring/removing tests for a breaking
> change like this? There are quite a few failing, would you change the logic
> to reflect the new type inference (or lack thereof, in this case)?
>
> For Index constructor, if you call it yourself you can pass the dtype
> param which will disable the type coercion. But, when the constructor is
> called from *inside* set_index(), it doesn't pass dtype param. The change
> I made there simply grabs the type of the col you're trying to create an
> index from, and passes that on to the constructor and uses the logic
> already in place.
>
> Still learning the code base, but I'm wondering if there's a way to do
> this with less overall impact than the way I did it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/30870?email_source=notifications&email_token=AAKAOISLMZCB6XOKWHH6HWDQ5U4OBA5CNFSM4KE7USY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI3G6YY#issuecomment-573992803>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAOIQYIG3QXRLHUHHMLY3Q5U4OBANCNFSM4KE7USYQ>
> .
>

jreback commented:
@trevorbye yeah I agree with @TomAugspurger what we need to do here is actually fix this. This will by-definition break *some* tests which will need to be updated, some because we are specifically testing that functionaility, and some by happen-stance (we happened to construct the expected value with an inferred type).

So as I said above, my preference here is to leave the Index constructor alone, and simply pass in the dtype of what we are setting (now this will have to be slightly adjusted, so create a function to do this), e.g. a int* column -> np.dtype64 (as we don't support anything but Int64Index).

trevorbye commented:
> @trevorbye yeah I agree with @TomAugspurger what we need to do here is actually fix this. This will by-definition break _some_ tests which will need to be updated, some because we are specifically testing that functionaility, and some by happen-stance (we happened to construct the expected value with an inferred type).
> 
> So as I said above, my preference here is to leave the Index constructor alone, and simply pass in the dtype of what we are setting (now this will have to be slightly adjusted, so create a function to do this), e.g. a int* column -> np.dtype64 (as we don't support anything but Int64Index).

@jreback thanks, I can refactor all those tests once we figure out the logic.

I think the change I made here is what you described; No change to Index constructor, I just passed it the current dtype. For reference, here is the main change I made, line 4266 in frame.py in my commit:

```python
if found:
    # get current dtype to preserve through index creation
    current_dtype = self.dtypes.get(col).type
```

This is then passed to `ensure_index_from_sequences(arrays, names, current_dtype)` (I created the new optional param here), which is then forwarded on to the Index constructor. I've done some debugging, and from what I'm seeing, the Index (and subclasses like Int64Index etc.) constructors will automatically handle unsupported index types if you attempt to create them. Here's an example:

```
In [1]: df = pd.DataFrame({'A': pd.Series([1, 2, 3], dtype=np.int32), 'B': [1, 2, 3]})

In [2]: print(df.dtypes.get('A').type)
Out [2]: <class 'numpy.int32'>

In [3]: print(df.set_index('A').index)
Out [3]: Int64Index([1, 2, 3], dtype='int64', name='A')
```

I ran the above code with my changes, and even though it's grabbing `np.int32` from the series type and using that as a param to the Index constructor dtype, it automatically detects that it's unsupported as an index type and coerces to int64.

So I don't think we need a new function to pre-convert these types before passing to Index constructors, as the logic is already there. Is this what you meant by a new function? Apologies if I'm going in circles, but to me it seems like the implementation we're discussing, and the implementation I made, are the same core logic, minus any edge cases or additional considerations I may be missing.

Thanks for your help!

trevorbye commented:
@jreback any feedback on my above comment? Want to see what you think before I go refactor a ton of tests

jreback commented:
will look soon; you have a lot of pickle files committed; remove those

Commit:
removing test pkl files

jreback commented on pandas/core/frame.py:
```diff
@@ -4317,7 +4321,7 @@ def set_index(
                 )
 
-        index = ensure_index_from_sequences(arrays, names)
+        index = ensure_index_from_sequences(arrays, names, current_dtype)
```
I think you *might* able to make a much more direct change.

line 4311, something like
```
col = frame[col]
arrays.append(Index(col, dtype=col.dtype)
names.append(col)
```

jreback replied:
and you don't need the rest of the changes

trevorbye replied:
yes good point, I'll work on this

trevorbye replied:
@jreback I tested this out by directly appending the Index like in your example, and it did not override the dtype. This looks to be because when the call stack gets inside `ensure_index_from_sequences` where it actually creates the index on the line `return Index(sequences[0], name=names)`, if you do it this way you essentially get `return Index(Index(col, col.dtype), name=names)`. In a sense, the inner constructor in this example has your custom dtype, but by not setting that param in the outer constructor, it just reverts to type inference.

I think we need changes more like what I initially attempted e.g. we need to directly set the dtype in the Index constructor inside of `ensure_index_from_sequences`.

jreback commented on pandas/core/frame.py:
```diff
@@ -4492,9 +4496,6 @@ class    max    type
 
         def _maybe_casted_values(index, labels=None):
             values = index._values
-            if not isinstance(index, (PeriodIndex, DatetimeIndex)):
```
this is ok to still remove

Commit:
hard reset

Commit:
pep fix

Commit:
black

Commit:
isort

Commit:
restoring actions file

trevorbye commented:
@jreback @TomAugspurger I have a git bug (likely on my end), but I'm unable to push `.github/workflows/assign.yml` to my fork because I'm getting an OAuth error that I don't have workflow scope. Ever encountered this before/know of an easy fix? I have it temporarily removed in my branch right now while I was working on getting my PR finished

trevorbye commented:
@jreback @TomAugspurger want to take another look? See my [comment](https://github.com/pandas-dev/pandas/pull/30870#discussion_r371485842) on your review, I tried your suggestion of doing `arrays.append(Index(col, dtype=col.dtype)`, but setting the dtype gets overridden by the outer constructor in that case.

There was some additional code I added as well to allow as many tests to pass as possible without altering them. I excluded datetime64 from this new logic, because there is a lot that breaks if you don't leave that as it is. Also added a try/catch for some edge cases that occur in some tests.

Commit:
whatsnew entry

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into dtype-patch

trevorbye commented:
@jreback @jorisvandenbossche @WillAyd @TomAugspurger hi guys, can anyone take a look/see if we can merge? Thanks for your help

TomAugspurger commented:
Focused on regressions from the 1.0 release right now. Should have a chance
to look next week.

On Fri, Jan 31, 2020 at 1:06 PM trevorbye <notifications@github.com> wrote:

> @jreback <https://github.com/jreback> @jorisvandenbossche
> <https://github.com/jorisvandenbossche> @WillAyd
> <https://github.com/WillAyd> @TomAugspurger
> <https://github.com/TomAugspurger> hi guys, can anyone take a look/see if
> we can merge? Thanks for your help
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/30870?email_source=notifications&email_token=AAKAOIT4VHU6W6B4AK54D2TRARZB3A5CNFSM4KE7USY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKPVQMI#issuecomment-580868145>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAOIVLXTXFG4I7M3YPZWTRARZB3ANCNFSM4KE7USYQ>
> .
>

WillAyd commented on pandas/tests/extension/base/groupby.py:
```diff
@@ -19,6 +19,7 @@ def test_grouping_grouper(self, data_for_grouping):
         tm.assert_numpy_array_equal(gr1.grouper, df.A.values)
         tm.assert_extension_array_equal(gr2.grouper, data_for_grouping)
 
+    @pytest.mark.skip(reason="logic change to stop coercing dtypes on set_index()")
```
So you are skipping these because the existing tests are incompatible with the new logic right? 

If that's correct then we should actually update the tests and not just skip

trevorbye replied:
Yes, I will refactor them. From what I could see looking through tests, it seemed to me like a lot of them are skipped rather than refactored. 

Would best practice be to try and always refactor? Thanks for your feedback

WillAyd replied:
This PR is rarer than others as it intentionally is breaking existing behavior; in such case yes would want to update existing tests and not just skip them

Also needs a little more communication in whatsnew around backwards compatibility but can leave that until later; I think would be helpful to see tests updated for reviewers to have a better understanding of scope of changes this would cause

trevorbye replied:
Okay, that make sense. I'll work on getting these tests updated so it is clearer what the scope of the change is. In general, these few I skipped were asserting two df's to be equal, which is true *except* of course for the index of each df were no longer equal, thus they failed.

jreback replied:
yes, you need to actually change the results

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -148,6 +148,8 @@ Indexing
 ^^^^^^^^
 - Bug in slicing on a :class:`DatetimeIndex` with a partial-timestamp dropping high-resolution indices near the end of a year, quarter, or month (:issue:`31064`)
 - Bug in :meth:`PeriodIndex.get_loc` treating higher-resolution strings differently from :meth:`PeriodIndex.get_value` (:issue:`31172`)
+- Bug in :meth:`DataFrame.set_index` not preserving column dtype when creating :class:`Index` from a single column (:issue:`30517`)
```
move this to a sub-section; show a typical example of before & after (see how this is done in the Enhancements section)

jreback commented on pandas/core/frame.py:
```diff
@@ -4320,6 +4320,16 @@ def set_index(
                 try:
                     found = col in self.columns
+                    if found:
+                        # get current dtype to preserve through index creation,
```
what does this all mean? why would we not always preserve?

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -5458,6 +5458,7 @@ def ensure_index_from_sequences(sequences, names=None):
     ----------
     sequences : sequence of sequences
     names : sequence of str
+    dtype : NumPy dtype
```
why is this a NumPy dtype? type dtype: Dtype

WillAyd commented:
@trevorbye is this still active? Can you address comments and fix merge conflicts?

WillAyd commented:
Closing as I think stale, but ping if you'd like to continue and can reopen
